File: create_ipk.py
#!/usr/bin/env python3
"""
create_ipk.py  —  Build a webOS .ipk package without the full SDK.
Requires only the Python 3 standard library.

Usage:
    python create_ipk.py [app-dir] [output-dir]

Defaults:
    app-dir    = current directory  (must contain appinfo.json)
    output-dir = ./build
"""

import os, sys, io, struct, tarfile, hashlib, json, shutil, time, gzip

APP_DIR  = sys.argv[1] if len(sys.argv) > 1 else '.'
OUT_DIR  = sys.argv[2] if len(sys.argv) > 2 else os.path.join(APP_DIR, 'build')

# ── Read appinfo.json ──────────────────────────────────────────────────────────
appinfo_path = os.path.join(APP_DIR, 'appinfo.json')
if not os.path.exists(appinfo_path):
    sys.exit('ERROR: appinfo.json not found in ' + APP_DIR)

with open(appinfo_path, 'r', encoding='utf-8') as fh:
    appinfo = json.load(fh)

APP_ID      = appinfo.get('id',      'com.streamdeck.app')
APP_VERSION = appinfo.get('version', '1.0.0')
APP_DIR_INSTALL = '/usr/palm/applications/' + APP_ID

os.makedirs(OUT_DIR, exist_ok=True)

# ── Files to package ──────────────────────────────────────────────────────────
SKIP_NAMES = {
    'build', '.git', '__pycache__',
    'create_ipk.py', 'generate_icons.js',
    'build.bat', 'build.sh', 'README.md',
    'package.bat'
}
SKIP_EXTS = {'.py', '.bat', '.sh', '.ipk'}

def collect_files(base_dir):
    result = []
    for root, dirs, files in os.walk(base_dir):
        # prune skipped directories
        dirs[:] = [d for d in dirs if d not in SKIP_NAMES]
        rel_root = os.path.relpath(root, base_dir)
        for fname in files:
            if fname in SKIP_NAMES:
                continue
            _, ext = os.path.splitext(fname)
            if ext in SKIP_EXTS:
                continue
            rel_path = os.path.join(rel_root, fname) if rel_root != '.' else fname
            abs_path = os.path.join(root, fname)
            result.append((rel_path.replace('\\', '/'), abs_path))
    return result

app_files = collect_files(APP_DIR)
if not app_files:
    sys.exit('ERROR: No app files found in ' + APP_DIR)

print('Packaging ' + str(len(app_files)) + ' file(s)…')

# ── Helper: create in-memory tar.gz ───────────────────────────────────────────
def make_targz(entries):
    """
    entries: list of (arcname, bytes_or_filepath)
    Returns: bytes of the .tar.gz
    """
    buf = io.BytesIO()
    with tarfile.open(fileobj=buf, mode='w:gz', compresslevel=9) as tar:
        for arcname, content in entries:
            if isinstance(content, (bytes, bytearray)):
                info = tarfile.TarInfo(name=arcname)
                info.size  = len(content)
                info.mtime = int(time.time())
                info.mode  = 0o644
                tar.addfile(info, io.BytesIO(content))
            else:
                # file path
                tar.add(content, arcname=arcname)
    return buf.getvalue()

# ── 1.  debian-binary ─────────────────────────────────────────────────────────
debian_binary = b'2.0\n'

# ── 2.  control.tar.gz ────────────────────────────────────────────────────────
# Calculate installed size (approximate, in KB)
installed_kb = sum(os.path.getsize(fp) for _, fp in app_files) // 1024 + 1

control_text = (
    'Package: {id}\n'
    'Version: {ver}\n'
    'Architecture: all\n'
    'Installed-Size: {size}\n'
    'Maintainer: StreamDeck\n'
    'Description: Stream Deck IPTV Application for LG webOS\n'
    'webOS-Package-Format-Version: 2\n'
    'App-Id: {id}\n'
    'App-Version: {ver}\n'
).format(id=APP_ID, ver=APP_VERSION, size=installed_kb)

control_gz = make_targz([('./control', control_text.encode('utf-8'))])

# ── 3.  data.tar.gz ───────────────────────────────────────────────────────────
data_entries = []
for rel, abs_path in app_files:
    arc = '.' + APP_DIR_INSTALL + '/' + rel
    data_entries.append((arc, abs_path))

data_gz = make_targz(data_entries)

# ── 4.  Assemble AR archive (.ipk = Debian .deb format) ──────────────────────
AR_MAGIC = b'!<arch>\n'

def ar_header(name, size):
    """Return a 60-byte AR file header."""
    name_b  = name.encode('ascii').ljust(16)[:16]
    mtime_b = str(int(time.time())).encode('ascii').ljust(12)[:12]
    uid_b   = b'0     '
    gid_b   = b'0     '
    mode_b  = b'100644  '
    size_b  = str(size).encode('ascii').ljust(10)[:10]
    magic_b = b'\x60\x0a'
    return name_b + mtime_b + uid_b + gid_b + mode_b + size_b + magic_b

def ar_entry(name, data):
    header = ar_header(name, len(data))
    # Data padded to even byte boundary
    pad    = b'\n' if len(data) % 2 else b''
    return header + data + pad

ipk_buf = (
    AR_MAGIC +
    ar_entry('debian-binary',   debian_binary) +
    ar_entry('control.tar.gz',  control_gz)    +
    ar_entry('data.tar.gz',     data_gz)
)

ipk_name = APP_ID + '_' + APP_VERSION + '_all.ipk'
ipk_path = os.path.join(OUT_DIR, ipk_name)

with open(ipk_path, 'wb') as fh:
    fh.write(ipk_buf)

size_kb = len(ipk_buf) // 1024
print()
print('IPK created:  ' + ipk_path)
print('Size:         ' + str(size_kb) + ' KB')
print()
print('Install on TV (Developer Mode must be active):')
print('  ares-install --device tv ' + ipk_path)
print('  OR copy to USB stick and install via Developer Mode app.')

File: create_ipk.js
/**
 * create_ipk.js  —  Build a webOS .ipk without the SDK or Python.
 * Requires Node.js >= 10 (only built-in modules: fs, path, zlib, child_process).
 *
 * Usage:
 *   node create_ipk.js [app-dir] [out-dir]
 */
"use strict";

var fs = require("fs");
var path = require("path");
var zlib = require("zlib");

var APP_DIR = process.argv[2] || __dirname;
var OUT_DIR = process.argv[3] || path.join(APP_DIR, "build");

// ── Read appinfo.json ─────────────────────────────────────────────────────────
var appinfoPath = path.join(APP_DIR, "appinfo.json");
if (!fs.existsSync(appinfoPath)) {
  process.exit("ERROR: appinfo.json not found in " + APP_DIR);
}
var appinfo = JSON.parse(fs.readFileSync(appinfoPath, "utf8"));
var APP_ID = appinfo.id || "com.streamdeck.app";
var APP_VERSION = appinfo.version || "1.0.0";
var INSTALL_DIR = "/usr/palm/applications/" + APP_ID;

if (!fs.existsSync(OUT_DIR)) {
  fs.mkdirSync(OUT_DIR, { recursive: true });
}

// ── Skip list ─────────────────────────────────────────────────────────────────
var SKIP_NAMES = [
  "build",
  "backup ipk",
  ".git",
  "__pycache__",
  "node_modules",
  "create_ipk.py",
  "create_ipk.js",
  "generate_icons.js",
  "build.bat",
  "build.sh",
  "README.md",
  "package.bat",
  "hls.min.js.bak.0.8.9"
];
var SKIP_EXTS = [".py", ".bat", ".sh", ".ipk", ".js.map"];

function collectFiles(base) {
  var result = [];
  function walk(dir) {
    fs.readdirSync(dir).forEach(function (name) {
      if (SKIP_NAMES.indexOf(name) !== -1) {
        return;
      }
      var full = path.join(dir, name);
      var stat = fs.statSync(full);
      if (stat.isDirectory()) {
        walk(full);
        return;
      }
      var ext = path.extname(name).toLowerCase();
      if (SKIP_EXTS.indexOf(ext) !== -1) {
        return;
      }
      var rel = path.relative(base, full).replace(/\\/g, "/");
      result.push({ rel: rel, abs: full, size: stat.size });
    });
  }
  walk(base);
  return result;
}

var files = collectFiles(APP_DIR);
if (!files.length) {
  process.exit("ERROR: No app files found in " + APP_DIR);
}

// Auto-strip trailing commas before ) from JS files — the formatter re-introduces
// them and Chromium 38 (webOS 3.x) throws SyntaxError on every build if not removed.
var JS_FILES_TO_FIX = ["app.js", "login.js", "polyfills.js"];
JS_FILES_TO_FIX.forEach(function (name) {
  var filePath = path.join(APP_DIR, name);
  if (!fs.existsSync(filePath)) return;
  var lines = fs.readFileSync(filePath, "utf8").split("\n");
  var count = 0;
  for (var i = 0; i < lines.length; i++) {
    if (/,\s*$/.test(lines[i])) {
      var j = i + 1;
      while (j < lines.length && /^\s*$/.test(lines[j])) j++;
      var nx = (lines[j] || "").trim();
      if (nx.charAt(0) === ")") {
        lines[i] = lines[i].replace(/,(\s*)$/, "$1");
        count++;
      }
    }
  }
  if (count > 0) {
    fs.writeFileSync(filePath, lines.join("\n"), "utf8");
    console.log("  ES5 fix: removed " + count + " trailing commas from " + name);
  }
});

console.log("Packaging " + files.length + " file(s)…");

// ── CRC-32 ────────────────────────────────────────────────────────────────────
var CRC_TABLE = (function () {
  var t = new Array(256);
  for (var n = 0; n < 256; n++) {
    var c = n;
    for (var k = 0; k < 8; k++) {
      c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    }
    t[n] = c;
  }
  return t;
})();
function crc32(buf) {
  var crc = 0xffffffff;
  for (var i = 0; i < buf.length; i++) {
    crc = CRC_TABLE[(crc ^ buf[i]) & 0xff] ^ (crc >>> 8);
  }
  return (crc ^ 0xffffffff) >>> 0;
}

// ── Minimal tar builder ───────────────────────────────────────────────────────
function padRight(s, len, ch) {
  ch = ch || " ";
  while (s.length < len) {
    s += ch;
  }
  return s.slice(0, len);
}
function padLeft(s, len, ch) {
  ch = ch || " ";
  while (s.length < len) {
    s = ch + s;
  }
  return s.slice(0, len);
}

function tarHeader(name, size, isDir) {
  var h = Buffer.alloc(512, 0);
  var nameBuf = Buffer.from(name, "utf8");
  // If name > 100 bytes, use GNU LongLink (simplified: truncate to 99)
  nameBuf.copy(h, 0, 0, Math.min(nameBuf.length, 99));
  // mode
  Buffer.from(padLeft(isDir ? "755" : "644", 7, "0") + " ").copy(h, 100);
  // uid, gid
  Buffer.from("0000000 ").copy(h, 108);
  Buffer.from("0000000 ").copy(h, 116);
  // size  (octal, 11 digits + space)
  Buffer.from(padLeft(size.toString(8), 11, "0") + " ").copy(h, 124);
  // mtime
  Buffer.from(padLeft(Math.floor(Date.now() / 1000).toString(8), 11, "0") + " ").copy(h, 136);
  // type
  h[156] = isDir ? 0x35 : 0x30; // '5' dir, '0' file
  // ustar magic
  Buffer.from("ustar  \0").copy(h, 257);

  // checksum placeholder
  Buffer.from("        ").copy(h, 148);
  var sum = 0;
  for (var i = 0; i < 512; i++) {
    sum += h[i];
  }
  Buffer.from(padLeft(sum.toString(8), 6, "0") + "\0 ").copy(h, 148);
  return h;
}

function makeTar(entries) {
  var chunks = [];
  entries.forEach(function (e) {
    if (e.dir) {
      chunks.push(tarHeader(e.name, 0, true));
      return;
    }
    var data = Buffer.isBuffer(e.data) ? e.data : Buffer.from(e.data);
    var header = tarHeader(e.name, data.length, false);
    chunks.push(header);
    chunks.push(data);
    // pad to 512-byte boundary
    var remainder = data.length % 512;
    if (remainder) {
      chunks.push(Buffer.alloc(512 - remainder, 0));
    }
  });
  // two zero 512-byte blocks at end
  chunks.push(Buffer.alloc(1024, 0));
  return Buffer.concat(chunks);
}

function makeTarGz(entries) {
  return zlib.gzipSync(makeTar(entries), { level: 9 });
}

// ── 1. debian-binary ─────────────────────────────────────────────────────────
var debianBinary = Buffer.from("2.0\n");

// ── 2. control.tar.gz ────────────────────────────────────────────────────────
var installedKb = files.reduce(function (s, f) {
  return s + f.size;
}, 0);

var controlText =
  "Package: " +
  APP_ID +
  "\n" +
  "Version: " +
  APP_VERSION +
  "\n" +
  "Section: misc\n" +
  "Priority: optional\n" +
  "Architecture: all\n" +
  "Installed-Size: " +
  installedKb +
  "\n" +
  "Maintainer: " +
  (appinfo.vendor || "Unknown") +
  "\n" +
  "Description: " +
  (appinfo.appDescription || appinfo.title || APP_ID) +
  "\n" +
  "webOS-Package-Format-Version: 2\n" +
  "webOS-Packager-Version: 3.2.5\n" +
  "App-Id: " +
  APP_ID +
  "\n" +
  "App-Version: " +
  APP_VERSION +
  "\n";

var controlGz = makeTarGz([{ name: "./control", data: controlText }]);

// ── 3. data.tar.gz ────────────────────────────────────────────────────────────
// webOS registers the package from usr/palm/packages/<id>/packageinfo.json;
// without it appinstalld reports a generic "-1 failed to install" error.
var packageInfo = {
  id: APP_ID,
  version: APP_VERSION,
  app: APP_ID
};

// Path layout mirrors ares-package: no "./" prefix, trailing "/" on directories.
var APP_ROOT = "usr/palm/applications/" + APP_ID;
var dataEntries = [
  { name: "usr/", dir: true },
  { name: "usr/palm/", dir: true },
  { name: "usr/palm/applications/", dir: true },
  { name: "usr/palm/packages/", dir: true },
  { name: APP_ROOT + "/", dir: true },
  { name: "usr/palm/packages/" + APP_ID + "/", dir: true }
];

files.forEach(function (f) {
  var parts = f.rel.split("/");
  var acc = APP_ROOT;
  for (var i = 0; i < parts.length - 1; i++) {
    acc += "/" + parts[i];
    var dirName = acc + "/";
    var already = dataEntries.some(function (e) {
      return e.dir && e.name === dirName;
    });
    if (!already) {
      dataEntries.push({ name: dirName, dir: true });
    }
  }
  dataEntries.push({
    name: APP_ROOT + "/" + f.rel,
    data: fs.readFileSync(f.abs)
  });
});

dataEntries.push({
  name: "usr/palm/packages/" + APP_ID + "/packageinfo.json",
  data: JSON.stringify(packageInfo, null, 2)
});

var dataGz = makeTarGz(dataEntries);

// ── 4. AR archive ─────────────────────────────────────────────────────────────
function arHeader(name, size) {
  var h = Buffer.alloc(60, 0x20);
  Buffer.from(name).copy(h, 0);
  Buffer.from(String(Math.floor(Date.now() / 1000))).copy(h, 16);
  Buffer.from("0").copy(h, 28); // uid
  Buffer.from("0").copy(h, 34); // gid
  Buffer.from("100644").copy(h, 40); // mode
  Buffer.from(String(size)).copy(h, 48);
  h[58] = 0x60;
  h[59] = 0x0a; // "`\n"
  return h;
}

function arEntry(name, data) {
  var header = arHeader(name, data.length);
  var pad = data.length % 2 ? Buffer.from([0x0a]) : Buffer.alloc(0);
  return Buffer.concat([header, data, pad]);
}

var ipkBuf = Buffer.concat([
  Buffer.from("!<arch>\n"),
  arEntry("debian-binary", debianBinary),
  arEntry("control.tar.gz", controlGz),
  arEntry("data.tar.gz", dataGz)
]);

var ipkName = APP_ID + "_" + APP_VERSION + "_all.ipk";
var ipkPath = path.join(OUT_DIR, ipkName);
fs.writeFileSync(ipkPath, ipkBuf);

console.log("");
console.log("IPK created : " + ipkPath);
console.log("Size        : " + Math.ceil(ipkBuf.length / 1024) + " KB");
console.log("");
console.log("Install on TV (Developer Mode active):");
console.log("  ares-install --device tv " + ipkPath);
